rust/src/dataset.rs

```diff
@@ -487,6 +487,35 @@ impl Dataset {
         Ok(counts.iter().sum())
     }
 
+    /// Count the number of rows in the dataset with filter.
+    pub async fn count_rows_with_filter(&self, expr: &str) -> Result<usize> {
+        // Open file to read metadata.
+        let count = stream::iter(self.get_fragments())
+            .map(|f| async move {
+                let mut scanner = f.scan();
+                let filter = scanner.filter(expr);
```

Comment:
Can we move this code into `Fragment::count_rows_with_filter()`? 
And within the dataset, just call

```
stream::iter(dataset.get_fragments()).map(|f| f.count_rows_with_filter(expr)).sum()
```

It could leave room for distributed counting as well, also better encapture.

Reply:
i don't think `scan.filter()` actually do column projection, does it?

For example for a dataset which schema like

```
<a: string, b: bytes, c: bool, d: int>
```

when you do `count_rows_by_filter("d > 10"), it should only read column `d` in this case.